### ISSUE TITLE
ModelCheckpoint score_function confusing to use

### ISSUE THREAD

**Issue by oteph** [pytorch/ignite#677]
I use `create_supervised_trainer` in combination with `ModelCheckpoint`.

```py    
checkpoint = ModelCheckpoint(
    f"./models/{start_time}",
    "resnet_50",
    score_name="loss",
    score_function=lambda engine: 1 - engine.state.output, # output contains the minibatch output
    n_saved=10,
    create_dir=True,
)
```

With supervised trainer in not straight forward to add a metric. So users only have access to the state.output, which only contains the minibatch. This causes models to be saved that based on the model performance on the last minibatch, not the whole epoch.

It might be a good idea to add a warning to the description of Modelcheckpoint that output contains the minibach output only. And/or add an example that shows how to use it in combination with a metric.

**Comment by vfdev-5** [comment]
@oteph thanks for the feedback! Could you please explain what would you like to achieve using `ModelCheckpoint` on the trainer ?

In general, `ModelCheckpoint` is used to create training checkpoints and save best models according to validation scores.
1) Create training checkpoints
```python

```
2) Save best models 
```python

```


> With supervised trainer in not straight forward to add a metric. So users only have access to the state.output, which only contains the minibatch. This causes models to be saved that based on the model performance on the last minibatch, not the whole epoch.

I assume that you are also aware of the fact that during the training, the model is changing and metrics computed during the training do not represent the the last performance of the model.
To add a running average metrics with ignite is simple, see https://pytorch.org/ignite/v0.2.1/metrics.html#ignite.metrics.RunningAverage
With this class you can trace running average of the loss and use the score function based on this metric.

Otherwise, as score function requires only engine, you can store your custom score in `engine.state` and provide it as score
```python
def score_function(engine):
    return engine.state.my_score
```


HTH

**Comment by oteph** [comment]
@vfdev-5 
> I assume that you are also aware of the fact that during the training, the model is changing and metrics computed during the training do not represent the the last performance of the model.

I was aware, but now that you point it out it makes sense that the `create_supervised_trainer ` does not take any metrics.

But this still seems problematic for the `ModelCheckpoint` `score_function` as either you can only access the output of the last minibatch, or a metric that does not represent the final model state.

**Comment by vfdev-5** [comment]
> But this still seems problematic for the `ModelCheckpoint` `score_function` as either you can only access the output of the last minibatch, or a metric that does not represent the final model state. 

@oteph or some custom value, as I said in the previous message...
It really depends on what would you need to store.

**Comment by oteph** [comment]
> > But this still seems problematic for the `ModelCheckpoint` `score_function` as either you can only access the output of the last minibatch, or a metric that does not represent the final model state.
> 
> @oteph or some custom value, as I said in the previous message...
> It really depends on what would you need to store.

@vfdev-5  alright what got me confused was the assumption that i cannot use the evaluator as EPOCH_COMPLETED gets triggered for the training and the validation set. So as long as i split the engines further and have one dedicated solely to the evaluation of the training batch, everything should work as expected. (as you mentioned in https://github.com/pytorch/ignite/issues/678#issuecomment-559789280)

Thanks for helping me out!

**Comment by vfdev-5** [comment]
@oteph hope to make it more clear. Also, do not hesitate to look at other example scripts and notebooks to see how ignite helps to get the code more flexible and refactored.

Feel free to close the issues if they are answered :)
